File: Editorial/174_Dungeon_Game/CSharp/1_Dynamic_Programming.cs
public class Solution {
    const int inf = int.MaxValue;
    int[,] dp;
    int rows, cols;

    public int GetMinHealth(int currCell, int nextRow, int nextCol) {
        if (nextRow >= this.rows || nextCol >= this.cols)
            return inf;
        int nextCell = this.dp[nextRow, nextCol];
        // hero needs at least 1 point to survive
        return Math.Max(1, nextCell - currCell);
    }

    public int CalculateMinimumHP(int[][] dungeon) {
        this.rows = dungeon.Length;
        this.cols = dungeon[0].Length;
        this.dp = new int[rows, cols];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                dp[i, j] = inf;
            }
        }

        int currCell, rightHealth, downHealth, nextHealth, minHealth;
        for (int row = this.rows - 1; row >= 0; --row) {
            for (int col = this.cols - 1; col >= 0; --col) {
                currCell = dungeon[row][col];

                rightHealth = GetMinHealth(currCell, row, col + 1);
                downHealth = GetMinHealth(currCell, row + 1, col);
                nextHealth = Math.Min(rightHealth, downHealth);

                if (nextHealth != inf) {
                    minHealth = nextHealth;
                } else {
                    minHealth = currCell >= 0 ? 1 : 1 - currCell;
                }

                this.dp[row, col] = minHealth;
            }
        }

        return this.dp[0, 0];
    }
}

File: Editorial/174_Dungeon_Game/C/1_Dynamic_Programming.c
int calculateMinimumHP(int** dungeon, int dungeonSize, int* dungeonColSize) {
    int rows = dungeonSize, cols = *dungeonColSize;
    int dp[rows][cols];
    memset(dp, 0, sizeof(dp));

    int get_min_health(int currCell, int nextRow, int nextCol) {
        if (nextRow >= rows || nextCol >= cols) {
            return INT_MAX;
        }
        int nextCell = dp[nextRow][nextCol];
        // hero needs at least 1 point to survive
        return nextCell - currCell >= 1 ? nextCell - currCell : 1;
    }

    for (int row = rows - 1; row >= 0; --row) {
        for (int col = cols - 1; col >= 0; --col) {
            int currCell = dungeon[row][col];

            int right_health = get_min_health(currCell, row, col + 1);
            int down_health = get_min_health(currCell, row + 1, col);
            int next_health =
                right_health < down_health ? right_health : down_health;

            int min_health;
            if (next_health != INT_MAX) {
                min_health = next_health;
            } else {
                min_health = currCell >= 0 ? 1 : 1 - currCell;
            }

            dp[row][col] = min_health;
        }
    }

    return dp[0][0];
}
